### PR TITLE
Add multi-character TRIM function

### PR DESCRIPTION
**Pull request by MrTeeett**
It's a reworked version of the old pull-request:
https://github.com/FirebirdSQL/firebird/pull/8004

### CODE REVIEW

File: src/dsql/parse.y [hunk 1]
```diff
@@ -703,6 +703,9 @@ using namespace Firebird;
 %token <metaNamePtr> CALL
 %token <metaNamePtr> FORMAT
 %token <metaNamePtr> NAMED_ARG_ASSIGN
+%token <metaNamePtr> BTRIM
```
Comment (asfernandes) [inline src/dsql/parse.y:706]: Reorder it alphabetically inside block `// tokens added for Firebird 6.0`, like the others.

File: src/dsql/parse.y [hunk 2]
```diff
@@ -4495,6 +4498,9 @@ keyword_or_column
 	| WINDOW
 	| WITHOUT
 	| CALL					// added in FB 6.0
```
Comment (asfernandes) [inline src/dsql/parse.y:4500]: Reorder it alphabetically inside block `// added in FB 6.0`, like the others.

File: src/dsql/parse.y [hunk 3]
```diff
@@ -8782,6 +8791,32 @@ trim_specification
...
 
+%type <valueExprNode> btrim_function
+btrim_function
+	: BTRIM '(' value ',' value ')'
+		{ $$ = newNode<TrimNode>(blr_trim_both, blr_trim_multi_characters, $3, $5); }
+	| BTRIM '(' value ')'
+		{ $$ = newNode<TrimNode>(blr_trim_both, blr_trim_spaces, $3); }
+	;
+
+
```
Comment (asfernandes) [inline src/dsql/parse.y:8802]:
```suggestion

```

File: src/dsql/parse.y [hunk 4]
```diff
@@ -8782,6 +8791,32 @@ trim_specification
...
+%type <valueExprNode> ltrim_function
+ltrim_function
+	: LTRIM '(' value ',' value ')'
+		{ $$ = newNode<TrimNode>(blr_trim_leading, blr_trim_multi_characters, $3, $5); }
+	| LTRIM '(' value ')'
+		{ $$ = newNode<TrimNode>(blr_trim_leading, blr_trim_spaces, $3); }
+	;
+
+
```
Comment (asfernandes) [inline src/dsql/parse.y:8811]:
```suggestion

```

File: doc/sql.extensions/README.trim [hunk 5]
```diff
@@ -18,6 +18,9 @@ Format:
 	<trim character> ::=
 		<value expression>
 
+	<multi-character trim function> ::=
+		{ BTRIM | LTRIM | RTRIM } <left paren> <trim source> [ <comma> <trim character> ] <right paren>
+
```
Comment (asfernandes) [inline doc/sql.extensions/README.trim:23]: Please document more explicitly difference between the new functions when using a `trim character` and specifiy that they work identically to previously function when `trim character` is not passed.

File: src/dsql/ExprNodes.cpp [hunk 6]
```diff
@@ -12831,6 +12834,52 @@ dsc* TrimNode::execute(thread_db* tdbb, Request* request) const
 				}
 			}
 		}
+
+		if (multi && (where == blr_trim_both || where == blr_trim_leading))
```
Comment (asfernandes) [inline src/dsql/ExprNodes.cpp:12838]: Use single `if (multi)` with others ifs inside it.

File: doc/sql.extensions/README.trim (hunk 5 above)
Comment (MrTeeett) [reply doc/sql.extensions/README.trim:23]: Corrected the comments, please check how else I need to finalize the documentation

File: src/common/ParserTokens.h [hunk 7]
```diff
@@ -507,6 +507,9 @@ PARSER_TOKEN(TOK_TRANSACTION, "TRANSACTION", true)
 PARSER_TOKEN(TOK_TRAPS, "TRAPS", true)
 PARSER_TOKEN(TOK_TRIGGER, "TRIGGER", false)
 PARSER_TOKEN(TOK_TRIM, "TRIM", false)
+PARSER_TOKEN(TOK_BTRIM, "BTRIM", false)
```
Comment (asfernandes) [inline src/common/ParserTokens.h:510]: Please put each entry in their alphabetically order.

File: src/dsql/ExprNodes.cpp [hunk 8]
```diff
@@ -12800,34 +12801,87 @@ dsc* TrimNode::execute(thread_db* tdbb, Request* request) const
...
+					if (memcmp(charactersCanonical.begin(),
+							   &valueCanonical[offsetTrail - charactersCanonicalLen],
+							   charactersCanonicalLen) != 0)
+					{
+						break;
+					}
+				}
+			}
+		}
+		if (multi)
```
Comment (asfernandes) [inline src/dsql/ExprNodes.cpp:12840]:
```suggestion
		else
```

**Pull request merged**